### DISCUSSION TITLE
Generating client from OpenAPI spec & data fetching

### DISCUSSION THREAD

**Discussion by MarcoMuellner** [nuxt/nuxt#17416] [General]
Hi guys,

i am currently trying to migrate my vue 3 project to nuxt3. There i use an [OpenAPI client generator](https://github.com/ferdikoomen/openapi-typescript-codegen) to generate my api client i use within the vue 3 project (using axios). This is incredibly convinient, as i get automated typing with typescript, and i don't have to adapt anything if i change the endpoints in the back (given they are called the same).

Now that i am trying to get into nuxt, as its abstractions and design choices are pretty cool. So i also stumbled upon the [data fetching](https://v3.nuxtjs.org/guide/features/data-fetching) part of the API. I would really like to use this (especially the refresh feature seems awesome), but i also really don't want to give up the client generation from the OpenAPI spec. 

Surely i am not the first one to want this, and there are solutions out there. However i couldn't really find one, is there any good advice on a problem like this?

**Comment by wobsoriano** [comment]
Would you be able to create a simple repro so I can try to play with it?

**Comment by MarcoMuellner** [comment]
Will defo do that, as soon as i can

**Reply by xeco8000** [reply]
Hello thanks for your contribution. I would also be interested in a solution. However, I would like to be able to fetch all the data from an API call on the server side and access it in all modules.

**Comment by oemer-aran** [comment]
I would really love to see an api generator for `$fetch` with nuxt :/ The generator is the only reason why i still use `axios` as openapi tools has  `typescript-axios` generator.

**Comment by huang-julien** [comment]
I'm also doing client generation from openapi, was planning to do that but didn't had time yet.

For now I just did a module which only bring type response for nitro endpoints.

**Comment by enkot** [comment]
You can try https://nuxt-open-fetch.vercel.app, it generates 100% typed OpenAPI fetch clients with zero overhead.
All ideas/suggestions to improve the API are welcome! :)

**Reply by enkot** [reply]
To clarify, it is a wrapper for `$fetch` and `useFetch` as mentioned in the docs (first sentence 🙂) - https://nuxt-open-fetch.vercel.app/setup/quick-start

**Comment by dbaranec** [comment]
@MarcoMuellner

any news, how to use 'useFetch' and 'useLazyFetch' with [OpenAPI client generator](https://github.com/ferdikoomen/openapi-typescript-codegen)  ?

**Comment by mrlubos** [comment]
@MarcoMuellner @wobsoriano @xeco8000 @oemer-aran @huang-julien @dbaranec if you're still working with Nuxt + OpenAPI, have a look at Hey API (https://heyapi.dev/). It's originally forked from the openapi-typescript-codegen library a few of you shared in this thread, but quite different today. I'm now working on a Nuxt client, you can try it and provide feedback in https://github.com/hey-api/openapi-ts/issues/998. Otherwise, you can wait until it's published on npm which should be soon.

@enkot if you plan to maintain your package, I'd consider moving it under the openapi-typescript project you're using for type generation. I am sure they'd be open to it!

UPDATE: Nuxt module is available on npm https://www.npmjs.com/package/@hey-api/nuxt

**Comment by Pentadome** [comment]
Hey guys, I also made a library for this. [nuxt-openAPI-wrapper](https://github.com/Pentadome/nuxt-openAPI-wrapper), unlike other solutions (i believe?), the wrapper still using Nuxt's own fetch client under the hood, so all the benefits of that client still work.

**Reply by mrlubos** [reply]
Just saw it today! I'd consider submitting it directly to openapi-typescript, you'll receive a lot more feedback and visibility.
